### PR TITLE
Fix multiplication of Triangular and Diagonal matrix

### PR DESCRIPTION
**Pull request by goggle**
This PR intends to fix some examples from https://github.com/JuliaLang/julia/issues/30094#issuecomment-440175887, more precisely the multiplication of a triangular matrix with a diagonal matrix and vice versa.

There are more broken examples involving triangular and diagonal matrices, but I first wanted to check if this is done right.

Here are some examples that still need to be addressed:
```julia
julia> using LinearAlgebra

julia> D = Diagonal([0.5, 0.5, 0.5]);

julia> T = LowerTriangular(ones(Int, (3,3)));

julia> T'*D
ERROR: InexactError: Int64(0.5)
Stacktrace:
 [1] Int64 at ./float.jl:709 [inlined]
 [2] convert at ./number.jl:7 [inlined]
 [3] setindex! at ./array.jl:798 [inlined]
 [4] setindex! at ./multidimensional.jl:490 [inlined]
 [5] macro expansion at ./broadcast.jl:909 [inlined]
 [6] macro expansion at ./simdloop.jl:77 [inlined]
 [7] copyto! at ./broadcast.jl:908 [inlined]
 [8] copyto! at ./broadcast.jl:863 [inlined]
 [9] materialize! at ./broadcast.jl:822 [inlined]
 [10] rmul!(::Array{Int64,2}, ::Diagonal{Float64,Array{Float64,1}}) at /home/alex/Projects/github/goggle/julia/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/diagonal.jl:177
 [11] rmul! at /home/alex/Projects/github/goggle/julia/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/diagonal.jl:187 [inlined]
 [12] *(::Adjoint{Int64,LowerTriangular{Int64,Array{Int64,2}}}, ::Diagonal{Float64,Array{Float64,1}}) at /home/alex/Projects/github/goggle/julia/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/diagonal.jl:219
 [13] top-level scope at REPL[4]:1

julia> transpose(T)*D
ERROR: InexactError: Int64(0.5)
Stacktrace:
 [1] Int64 at ./float.jl:709 [inlined]
 [2] convert at ./number.jl:7 [inlined]
 [3] setindex! at ./array.jl:798 [inlined]
 [4] setindex! at ./multidimensional.jl:490 [inlined]
 [5] macro expansion at ./broadcast.jl:909 [inlined]
 [6] macro expansion at ./simdloop.jl:77 [inlined]
 [7] copyto! at ./broadcast.jl:908 [inlined]
 [8] copyto! at ./broadcast.jl:863 [inlined]
 [9] materialize! at ./broadcast.jl:822 [inlined]
 [10] rmul!(::Array{Int64,2}, ::Diagonal{Float64,Array{Float64,1}}) at /home/alex/Projects/github/goggle/julia/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/diagonal.jl:177
 [11] rmul! at /home/alex/Projects/github/goggle/julia/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/diagonal.jl:187 [inlined]
 [12] *(::Transpose{Int64,LowerTriangular{Int64,Array{Int64,2}}}, ::Diagonal{Float64,Array{Float64,1}}) at /home/alex/Projects/github/goggle/julia/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/diagonal.jl:228
 [13] top-level scope at REPL[5]:1

julia> D*T'
ERROR: InexactError: Int64(0.5)
Stacktrace:
 [1] Int64 at ./float.jl:709 [inlined]
 [2] convert at ./number.jl:7 [inlined]
 [3] setindex! at ./array.jl:798 [inlined]
 [4] setindex! at ./multidimensional.jl:490 [inlined]
 [5] macro expansion at ./broadcast.jl:909 [inlined]
 [6] macro expansion at ./simdloop.jl:77 [inlined]
 [7] copyto! at ./broadcast.jl:908 [inlined]
 [8] copyto! at ./broadcast.jl:863 [inlined]
 [9] materialize! at ./broadcast.jl:822 [inlined]
 [10] lmul!(::Diagonal{Float64,Array{Float64,1}}, ::Array{Int64,2}) at /home/alex/Projects/github/goggle/julia/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/diagonal.jl:183
 [11] *(::Diagonal{Float64,Array{Float64,1}}, ::Adjoint{Int64,LowerTriangular{Int64,Array{Int64,2}}}) at /home/alex/Projects/github/goggle/julia/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/diagonal.jl:237
 [12] top-level scope at REPL[6]:1

julia> D*transpose(T)
ERROR: InexactError: Int64(0.5)
Stacktrace:
 [1] Int64 at ./float.jl:709 [inlined]
 [2] convert at ./number.jl:7 [inlined]
 [3] setindex! at ./array.jl:798 [inlined]
 [4] setindex!(::UpperTriangular{Int64,Array{Int64,2}}, ::Float64, ::Int64, ::Int64) at /home/alex/Projects/github/goggle/julia/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/triangular.jl:229
 [5] macro expansion at ./abstractarray.jl:1104 [inlined]
 [6] macro expansion at ./simdloop.jl:77 [inlined]
 [7] copyto! at ./broadcast.jl:908 [inlined]
 [8] copyto!(::UpperTriangular{Int64,Array{Int64,2}}, ::Base.Broadcast.Broadcasted{LinearAlgebra.StructuredMatrixStyle{UpperTriangular{Int64,Array{Int64,2}}},Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(*),Tuple{Array{Float64,1},UpperTriangular{Int64,Array{Int64,2}}}}) at /home/alex/Projects/github/goggle/julia/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/structuredbroadcast.jl:175
 [9] *(::Diagonal{Float64,Array{Float64,1}}, ::Transpose{Int64,LowerTriangular{Int64,Array{Int64,2}}}) at ./broadcast.jl:822
 [10] top-level scope at REPL[8]:1
```

And in this example, I think we should expect in both cases a `UpperTriangular` as a result, right?
```julia
julia> T = LowerTriangular(ones(Float64, (3,3)));

julia> T'*D
3×3 UpperTriangular{Float64,Array{Float64,2}}:
 0.5  0.5  0.5
  ⋅   0.5  0.5
  ⋅    ⋅   0.5

julia> D*T'
3×3 Array{Float64,2}:
 0.5  0.5  0.5
 0.0  0.5  0.5
 0.0  0.0  0.5
```

### CODE REVIEW

**Comment by andreasnoack** [comment]
The approach looks good to me.

**Comment by goggle** [comment]
I have now adopted the remaining multiplication methods in a similar fashion.

There are still some issues:
```julia
julia> using LinearAlgebra

julia> D = Diagonal([0.5, 0.5, 0.5]);

julia> L = LowerTriangular(ones(Int, (3,3)));

julia> L'*D
3×3 Array{Float64,2}:
 0.5  0.5  0.5
 0.0  0.5  0.5
 0.0  0.0  0.5

```

`L'*D` should be an `UpperTriangular`, right?

This would be solved, if `similar(L')` would return a `UpperTriangular`, which it currently doesn't:
```julia
julia> typeof(similar(L'))
Array{Int64,2}
```
So, should we add such `similar` methods? If so, where is the right place to do this?

**Comment by andreasnoack** [comment]
> So, should we add such similar methods? If so, where is the right place to do this?

Yes. I think it would make sense with such methods. I would think they should go in triangular.jl, if possible.

File: stdlib/LinearAlgebra/src/triangular.jl [hunk 1]
```diff
@@ -50,6 +50,20 @@ for t in (:LowerTriangular, :UnitLowerTriangular, :UpperTriangular,
...
+            LowerTriangular => UpperTriangular,
+            UpperTriangular => LowerTriangular,
+            UnitLowerTriangular => UnitUpperTriangular,
+            UnitUpperTriangular => UnitLowerTriangular,
+        )
+        similar(A::Adjoint{TI,TV}, ::Type{T}) where{T,TI,TV<:$t} = adjtypes[$t](similar(parent(parent(A)), T))
+        similar(A::Transpose{TI,TV}, ::Type{T}) where{T,TI,TV<:$t} = adjtypes[$t](similar(parent(parent(A)), T))
+    end
+end
+
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/triangular.jl]: If I'm not mistaken, then writing this out explicitly yields 8 lines of clear code, as opposed to, well, what you have. 😛

Comment (goggle) [reply stdlib/LinearAlgebra/src/triangular.jl]: That's correct, I did not think about this...

**Comment by goggle** [comment]
I have modified the code acccording to @dkarrasch's suggestion, so this PR is ready to be reviewed.

**Pull request merged**